File: code/framework-comparison/markup-renderer/markup_renderer.go
package main

import (
	"fmt"
	"io/ioutil"
	"regexp"
	"strings"
)

var markupTags = map[*regexp.Regexp]string{
	regexp.MustCompile(`\[title\](.*?)\[/title\]`):       "\033[33;1m$1\033[0m", // Yellow bold text
	regexp.MustCompile(`\[emphasis\](.*?)\[/emphasis\]`): "\033[31;1m$1\033[0m", // Red bold text
}

func renderMarkup(text string) string {
	for tag, replacement := range markupTags {
		text = tag.ReplaceAllString(text, replacement)
	}
	return text
}

func main() {
	inputFilename := "input.txt"

	content, err := ioutil.ReadFile(inputFilename)
	if err != nil {
		fmt.Printf("File '%s' not found.\n", inputFilename)
		return
	}

	renderedContent := renderMarkup(string(content))
	fmt.Println(renderedContent)
}

File: code/framework-comparison/markup-renderer/markup_renderer.swift
import Foundation

func renderMarkup(text: String) -> String {
    let markupTags = [
        ("\\[title\\](.*?)\\[/title\\]", "\u{001B}[33;1m$1\u{001B}[0m"),   // Yellow bold text
        ("\\[emphasis\\](.*?)\\[/emphasis\\]", "\u{001B}[31;1m$1\u{001B}[0m") // Red bold text
    ]
    
    var result = text
    for (tag, replacement) in markupTags {
        if let regex = try? NSRegularExpression(pattern: tag) {
            let range = NSRange(location: 0, length: result.utf16.count)
            result = regex.stringByReplacingMatches(in: result, options: [], range: range, withTemplate: replacement)
        }
    }
    return result
}

let inputFilename = "input.txt"
do {
    let content = try String(contentsOfFile: inputFilename)
    let renderedContent = renderMarkup(text: content)
    print(renderedContent)
} catch {
    print("File '\(inputFilename)' not found.")
}
